mongo/src/main/java/com/redhat/lightblue/mongo/crud/IterateAndUpdate.java

```diff
@@ -186,6 +200,16 @@ public void update(CRUDOperationContext ctx,
                 measure.begin("ctx.addDocument");
                 DocTranslator.TranslatedDoc translatedDoc=translator.toJson(document);
                 DocCtx doc=new DocCtx(translatedDoc.doc,translatedDoc.rmd);
+                if (memoryMonitor != null) {
+                    // if memory threshold is exceeded, this will throw an Error
+                    memoryMonitor.apply(doc);
```

Comment:
> On a separate note, I think there are other places in this algorithm the document is copied: at least when update starts, we copy the document again, and possibly also when the update is projected it may copy it as well (though I didn't read through that code well enough to see if it was just a pass-through-view upon the original or if a proper copy, though there is definitely non-zero memory overhead in either case). Should we monitor those too or is this approximation enough, given we have a separate threshold for writes?

Coming up with DocCtx size is tricky. There is originalDocument, outputDocument (projected) and  updatedDocument, all populated and re-populated at different stages of update operation processing. I included the copies created when updates starts (which you pointed out). Projection - as far as I can see - is only creating json "containers" (ArrayNode/ObjectNode) and keeps references to actual fields and values. To be honest, after days of looking at this I'm still unsure on how many document copies we store in memory at peak. It's not just the update that creates copies, it's also HookManager and I really don't know why it's not just using the copies in DocCtx'es.

At this point, the estimate sums up root, originalDocument and updatedDocument from DocCtx and pre and post copies created during hook queuing. What is not covered by the estimate is that updatedDocument is modified (by the update operation) and outputDocument created by projector.